tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/EncryptedEmailFastViewGetMethod.scala

```diff
@@ -0,0 +1,128 @@
+    computeResponse(request, mailboxSession)
+      .map(response => Invocation(
+        methodName = methodName,
+        arguments = Arguments(EncryptedEmailSerializer.serializeEncryptedEmailGetResponse(response).as[JsObject]),
+        methodCallId = invocation.invocation.methodCallId
+      ))
+      .map(InvocationWithContext(_, invocation.processingContext))
+
+  override def getRequest(mailboxSession: MailboxSession, invocation: Invocation): Either[Exception, EncryptedEmailGetRequest] =
+    EncryptedEmailSerializer.deserializeEncryptedEmailGetRequest(invocation.arguments.value) match {
+      case JsSuccess(emailGetRequest, _) => Right(emailGetRequest)
+      case errors: JsError => Left(new IllegalArgumentException(ResponseSerializer.serialize(errors).toString))
+    }
+
+  private def computeResponse(request: EncryptedEmailGetRequest,
+                              mailboxSession: MailboxSession): SMono[EncryptedEmailGetResponse] =
+    computeResults(request, mailboxSession)
+      .map(result => EncryptedEmailGetResponse(
+        accountId = request.accountId,
+        state = UuidState.INSTANCE,
+        list = result.list,
+        notFound = result.notFound
+      ))
+
+  private def computeResults(request: EncryptedEmailGetRequest,
+                             mailboxSession: MailboxSession): SMono[EncryptedEmailFastViewResults] = {
+    val parsedIds: List[Either[(UnparsedEmailId, IllegalArgumentException), MessageId]] = request.ids.value
+      .map(asMessageId)
+
+    val messagesIds: List[MessageId] = parsedIds.flatMap({
+      case Left(_) => None
+      case Right(messageId) => Some(messageId)
+    })
+
+    val parsingErrors: SFlux[EncryptedEmailFastViewResults] = SFlux.fromIterable(parsedIds.flatMap({
+      case Left((id, _)) =>
+        Some(EncryptedEmailFastViewResults.notFound(id))
+      case Right(_) => None
+    }))
+
+    SFlux.merge(Seq(parsingErrors, retrieveEmails(messagesIds, mailboxSession)))
+      .reduce(EncryptedEmailFastViewResults.empty())(EncryptedEmailFastViewResults.merge)
+  }
+
+  private def retrieveEmails(ids: Seq[MessageId], mailboxSession: MailboxSession): SFlux[EncryptedEmailFastViewResults] =
+    SFlux.fromIterable(ids)
+      .flatMap(messageId => retrieveEmail(messageId, mailboxSession))
+
+  private def retrieveEmail(ids: MessageId, mailboxSession: MailboxSession): SMono[EncryptedEmailFastViewResults] =
+    encryptedEmailFastViewReader.read(ids, mailboxSession)
+      .map(messageIdAndFastView => EncryptedEmailFastViewResults.list(messageIdAndFastView._1, messageIdAndFastView._2))
+      .onErrorResume {
+        case messageNotFound: MessageNotFoundException => SMono.just(EncryptedEmailFastViewResults.notFound(messageNotFound.messageId))
+        case messageNotFound: com.linagora.tmail.encrypted.MessageNotFoundException => SMono.just(EncryptedEmailFastViewResults.notFound(messageNotFound.messageId))
```

Comment:
Why do we have 2 different exceptions here?

Reply:
I want to separate two case 
- Message not found in GPG presentation 
- Message not found in MessageIdManager

Reply:
The end result is the same right? The encrypted preview cannot be found...